myst_nb/glue.py

```diff
@@ -0,0 +1,234 @@
+                if key == this_key:
+                    bundle = output["data"]
+                    bundle = {this_key: val for key, val in bundle.items()}
+                    outputs.append(bundle)
+    if len(outputs) == 0:
+        print(f"Did not find key {this_key} in notebook {path_ntbk}")
+        return
+    if len(outputs) > 1:
+        print(f"Multiple variables found for key: {this_key}. Returning first value.")
+    return outputs[0]
+
+
+def find_all_keys(ntbk, keys=None, path=None, logger=None):
+    """Find all `glue` keys in a notebook and return a dictionary with key: outputs."""
+    if keys is None:
+        keys = {}
+
+    for cell in ntbk.cells:
+        if cell.cell_type != "code":
+            continue
+
+        for output in cell["outputs"]:
+            meta = output.get("metadata", {})
+            if "scrapbook" in meta:
+                this_key = meta["scrapbook"]["name"]
+                if this_key in keys:
+                    msg = f"Over-writing pre-existing glue key: `{this_key}`"
+                    if logger is None:
+                        print(msg)
+                    else:
+                        logger.warning(msg, location=(path, None))
+                keys[this_key] = output
+    return keys
+
+
+# Role and directive for pasting
+class Paste(SphinxDirective):
+    required_arguments = 1
+    final_argument_whitespace = True
+    has_content = False
+
+    option_spec = {"id": directives.unchanged}
+
+    def run(self):
+        # TODO: Figure out how to report cell number in the location
+        #       currently, line numbers in ipynb files are not reliable
+        path, lineno = self.state_machine.get_source_and_line(self.lineno)
+        # Remove line number if we have a notebook because it is unreliable
+        if path.endswith(".ipynb"):
+            lineno = None
+        # Remove the suffix from path so its suffix is printed properly in logs
+        path = str(Path(path).with_suffix(""))
+        return [PasteNode(self.arguments[0], "directive", location=(path, lineno))]
+
+
+def paste_role(name, rawtext, text, lineno, inliner, options={}, content=[]):
+    path = inliner.document.current_source
+    # Remove line number if we have a notebook because it is unreliable
+    if path.endswith(".ipynb"):
+        lineno = None
+    path = str(Path(path).with_suffix(""))
+    return [PasteNode(text, "role", location=(path, lineno))], []
+
+
+# Transform to replace nodes with outputs
+class PasteNode(nodes.container):
+    """Represent a MimeBundle in the Sphinx AST, to be transformed later."""
+
+    def __init__(self, key, kind, location=None, rawsource="", *children, **attributes):
+        self.key = key
+        self.kind = kind
+        self.location = location
+        super().__init__("", **attributes)
+
+
+class PasteNodesToDocutils(SphinxTransform):
+    """Use the builder context to transform a CellOutputNode into Sphinx nodes."""
+
+    default_priority = 700
```

Comment:
We should add a comment of justification for why 700 was chosen, and ditto for `CellOutputsToNodes`. Also should one be a higher priority than the other?

Reply:
Ummmm...that makes a very strong assumption that I have logical reasoning behind my decisions...

But I did a [quick wikipedia search](https://en.wikipedia.org/wiki/700_(number)) and this is why I've decided that the number is 700:

> 700 (seven hundred) is the natural number following 699 and preceding 701. It is the sum of four consecutive primes (167 + 173 + 179 + 181).

Reply:
😆

Reply:
Yeh `PasteNodesToDocutils` needs to be run before `CellOutputsToNodes` (see 
72a5595)